### PR TITLE
New kube-proxy configuration API

### LINKED ISSUE #1409: KubeOne improve support for MetalLB in L2 mode (Kubeproxy IPVS strict ARP)
Similar Issue as for KKP: https://github.com/kubermatic/kubermatic/issues/7309

KubeOne needs an issue option to set the stirct ARP mode over multiple release cycle of KubeOne. Otherwise MetalLB could have problems at customer envrionments.

Two options:
- set strict ARP as default
- provide to configure strict arp as an option by a kubeone option

### PR DESCRIPTION
**Pull request by kron4eg**
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1409 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
New kube-proxy configuration API
```

### CODE REVIEW

File: pkg/apis/kubeone/types.go [hunk 1]
```diff
@@ -255,7 +255,36 @@ type ClusterNetworkConfig struct {
...
 	CNI *CNI `json:"cni,omitempty"`
-}
+	// KubeProxy
+	KubeProxy *KubeProxyConfig `json:"kubeProxy,omitempty"`
+}
+
+type KubeProxyConfig struct {
+	IPVS     *IPVSConfig `json:"ipvs"`
+	IPTables *IPTables   `json:"iptables"`
+}
```
Comment (xmudrii) [inline pkg/apis/kubeone/types.go:265]: Can you please comment this struct and its fields?

File: pkg/apis/kubeone/types.go [hunk 2]
```diff
@@ -255,7 +255,36 @@ type ClusterNetworkConfig struct {
...
+
+type IPVSConfig struct {
```
Comment (xmudrii) [inline pkg/apis/kubeone/types.go:267]: Can you ensure that comments in this struct are valid Go comments?

File: pkg/apis/kubeone/validation/validation.go [hunk 3]
```diff
@@ -233,6 +233,29 @@ func ValidateClusterNetworkConfig(c kubeone.ClusterNetworkConfig, fldPath *field
 	if c.CNI != nil {
 		allErrs = append(allErrs, ValidateCNI(c.CNI, fldPath.Child("cni"))...)
 	}
+	// allErrs = append(allErrs, ValidateClusterNetworkConfig(c.ClusterNetwork, field.NewPath("clusterNetwork"))...)
```
Comment (xmudrii) [inline pkg/apis/kubeone/validation/validation.go:236]: Why is this line commented?

File: pkg/apis/kubeone/types.go [hunk 4]
```diff
@@ -255,7 +255,36 @@ type ClusterNetworkConfig struct {
...
+	// ipvs scheduler
+	Scheduler string `json:"scheduler"`
```
Comment (xmudrii) [inline pkg/apis/kubeone/types.go:269]: Should we add some validation for this field?

File: pkg/apis/kubeone/types.go [hunk 5]
```diff
@@ -255,7 +255,36 @@ type ClusterNetworkConfig struct {
...
+	// excludeCIDRs is a list of CIDR's which the ipvs proxier should not touch
+	// when cleaning up ipvs services.
+	ExcludeCIDRs []string `json:"excludeCIDRs"`
```
Comment (xmudrii) [inline pkg/apis/kubeone/types.go:272]: Similar to the comment above, we could add some validation for this field as well.

File: pkg/templates/kubeadm/v1beta2/kubeadm.go [hunk 6]
```diff
@@ -367,3 +373,36 @@ func newNodeRegistration(s *state.State, host kubeoneapi.HostConfig) kubeadmv1be
...
 }
+
+func kubeProxyConfiguration(s *state.State) *kubeproxyv1alpha1.KubeProxyConfiguration {
+	kubeProxyConfig := &kubeproxyv1alpha1.KubeProxyConfiguration{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "KubeProxyConfiguration",
+			APIVersion: "kubeproxy.config.k8s.io/v1alpha1",
+		},
+		ClusterCIDR: s.Cluster.ClusterNetwork.PodSubnet,
+		BindAddress: "0.0.0.0",
```
Comment (xmudrii) [inline pkg/templates/kubeadm/v1beta2/kubeadm.go:384]: We could leave this out since it's the default value.

**Review by xmudrii — CHANGES_REQUESTED** [review]
Please make sure to update the example config as well.

File: pkg/apis/kubeone/types.go (hunk 4 above)
Comment (kron4eg) [reply pkg/apis/kubeone/types.go:269]: I'm not sure, original KubeProxyConfiguration doesn't have it. And default empty would work as well.

File: pkg/apis/kubeone/types.go (hunk 5 above)
Comment (kron4eg) [reply pkg/apis/kubeone/types.go:272]: Nether this one have an upstream validations.

**Review by xmudrii — APPROVED** [review]

**Pull request merged**